db.go

```diff
@@ -676,7 +676,43 @@ const (
 // Sync syncs database content to disk. This function provides
 // more control to user to sync data whenever required.
 func (db *DB) Sync() error {
-	return db.vlog.sync()
+	/**
+	Make an attempt to sync both the logs, the active memtable's WAL and the vLog (1847).
+	Cases:
+	- All_ok			:: If both the logs sync successfully.
+
+	- Entry_Lost		:: If an entry with a value pointer was present in the active memtable's WAL,
+						:: and the WAL was synced but there was an error in syncing the vLog.
+						:: The entry will be considered lost and this case will need to be handled during recovery.
+
+	- Entries_Lost		:: If there were errors in syncing both the logs, multiple entries would be lost.
+
+	- Entries_Lost      :: If the active memtable's WAL is not synced but the vLog is synced, it will
+						:: result in entries being lost because recovery of the active memtable is done from its WAL.
+						:: Check `UpdateSkipList` in memtable.go.
+
+	- Nothing_lost		:: If an entry with its value was present in the active memtable's WAL, and the WAL was synced,
+						:: but there was an error in syncing the vLog.
+						:: Nothing is lost for this very specific entry because the entry is completely present in the memtable's WAL.
+
+	- Partially_lost    :: If entries were written partially in either of the logs, the logs will be truncated during recovery.
+						:: As a result of truncation, some entries might be lost. Assume that 4KB of data is to be synced and
+					    :: invoking `Sync` results only in syncing 3KB of data and then the machine shuts down or the disk failure happens,
+	                    :: this will result in partial writes. [[This case needs verification]]
+	*/
+
+	var err error
+
+	memtableSyncError := db.mt.SyncWAL()
+	if memtableSyncError != nil {
+		err = errors.Wrap(memtableSyncError, "sync failed for the active memtable WAL")
+	}
+
+	vLogSyncError := db.vlog.sync()
+	if vLogSyncError != nil {
+		err = errors.Wrap(vLogSyncError, "sync failed for the vLog")
```

Comment:
does it make sense to combine this error with `memtableSyncError`?

Reply:
The idea is to perform a sync of vlog even if there is an error in syncing active memtable WAL. With this approach, I can return only one error.. hence, it is combined with memtableSyncError. 

We have very limited options:

1) If the sync of active memtable fails, do not perform vlog sync. So, no need to combine errors
2) Perform a sync of both the logs, combine errors
3) Perform a sync of both the logs, return a custom error struct

Reply:
option 2 is the best. We are still on Go 1.19, on Go 1.20 we could have used multi error.

Reply:
Yes, I originally used multi error but realized badger is not on golang 1.20. So, used `errors.Wrap` and combined both the errors.

Reply:
No, you are right. Corrected the code and added tests to checking error combination.